content/advent-2015/test2doc-generate-your-api-docs.md

```diff
@@ -0,0 +1,294 @@
+
+Enter *Go doc*.<br>
+We can find descriptions of an HTTP handler in its Go doc string, eg:
+```go
+// GetWidget retrieves a single Widget
+func GetWidget(w http.ResponseWriter, req *http.Request) {
+    // ...
+}
+```
+<br>
+And thus began **test2doc** - automatically generate complete API 
+documentation from your existing Go unit tests + Go doc strings.
+
+
+## Example
+
+Given an HTTP handler func, eg.:
+
+```go
+// GetWidget retrieves a single Widget
+func GetWidget(w http.ResponseWriter, req *http.Request) {
+    // ...
+}
+```
+
+And a test for this handler func, eg.:
+
+```go
+func TestGetWidget(t *testing.T) {
+    urlPath := fmt.Sprintf("/widgets/%d", 2)
+
+    resp, err := http.Get(server.URL + urlPath)
+    // assert all the things...
+}
+```
+
+Test2doc will automatically generate markdown documentation for this 
+endpoint in the [API 
+Blueprint](https://github.com/apiaryio/api-blueprint/blob/master/API%20B
+lueprint%20Specification.md) format as your tests run, like so:
+
+```
+# Group widgets
+
+## /widgets/{id}
+
++ Parameters
+    + id: `2` (number)
+
+### Get Widget [GET]
+retrieves a single Widget
+
++ Response 200 
+
+    + Body
+
+            {
+                "Id": 2,
+                "Name": "Pencil",
+                "Role": "Utensil"
+            }        
+```
+
+Which you can then parse and host w/ 
+[Apiary.io](http://docs.testingit.apiary.io/#): 
+![screenshot](/postimages/advent-2015/test2doc-widgets-api.jpg)
+
+Or use a custom parser and host yourself.
+
+## Getting started with test2doc
+A big goal of mine as I was writing this project was to limit the 
+amount of additional code needed to get test2doc up and running.
+
+Requirements:
+
+1. You must have a `TestMain` for each of the packages you wish to 
+document.
+2. All of the tests in your package must share a single `test.Server` 
+instance (from test2doc/test package)
+
+
+### 3 Code Additions
+
+3 additions, and only to your testing code:
+
+```go
+package widgets_test
+
+import (
+	"testing"
+
+	"github.com/adams-sarah/test2doc/test"
+)
+
+var server *test.Server
+
+func TestMain(m *testing.M) {
+	// 1. Tell test2doc how to get URL vars out of your HTTP requests
+	//
+	//    The 'URLVarExtractor' function must have the following signature:
+	//      func(req *http.Request) map[string]string
+	//      where the returned map is of the form map[key]value
+	test.RegisterURLVarExtractor(myURLVarExtractorFn)
+
+
+	// 2. You must use test2doc/test's wrapped httptest.Server instead of
+	//    the raw httptest.Server, so that test2doc can listen to and
+	//    record requests & responses.
+	//
+	//    NewServer takes your HTTP handler as an argument
+	server, err := test.NewServer(router)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// .. then run your tests as usual
+	exitCode := m.Run()
+
+
+	// 3. Finally, you must tell the wrapped server when you are done testing
+	//    so that the buffer can be flushed to an apib doc file
```

Comment:
hm, yea. "apib" is the extension for an API Blueprint file.
How about "...flushed to an API Blueprint doc file"?